app/controllers/api/v2/registration_controller.rb

```diff
@@ -0,0 +1,58 @@
+module Api
+  module V2
+    class RegistrationController < V2::BaseController
+      include Api::Version2
+      include Foreman::Controller::Registration
+      include Foreman::Controller::Parameters::Host
+
+      rescue_from ActiveRecord::RecordNotFound do |error|
+        logger.info "#{error.message} (#{error.class})"
+        not_found(error.message)
+      end
+
+      api :GET, '/register', N_('Render Global registration template')
+      param :organization_id, :number, desc: N_("ID of the Organization to register the host in.")
+      param :location_id, :number, desc: N_("ID of the Location to register the host in.")
+      param :hostgroup_id, :number, desc: N_("ID of the Host group to register the host in.")
+      param :operatingsystem_id, :number, desc: N_("ID of the Operating System to register the host in.")
+      def global
+        find_global_registration
+
+        unless @provisioning_template
+          not_found _('Global Registration Template with name %s defined via default_global_registration_item Setting not found, please configure the existing template name first') % Setting[:default_global_registration_item]
+          return
+        end
+
+        render plain: @provisioning_template.render(variables: @global_registration_vars).html_safe
+      end
+
+      api :POST, "/register", N_("Find or create a host and render the Host registration template")
+      param_group :host, Api::V2::HostsController
+      def host
+        begin
+          prepare_host
+          @template = @host.registration_template
+        rescue ::Foreman::Exception => e
+          render_error(e.message, status: :unprocessable_entity)
+          return
+        end
+
+        unless @template
+          not_found N_("Unable to find registration template for host %{host} running %{os}") % { host: @host.name, os: @host.operatingsystem }
+          return
+        end
+
+        @host.setBuild
+        safe_render(@template)
+      end
+
+      private
+
+      def prepare_host
+        @host = Host.find_or_initialize_by(name: host_params('host')['name'])
```

Comment:
Does it make sense to use the name or fqdn here? Can/should we also support finding the host via the puppet certname?

Reply:
I believe that is being extracted to a plugin, which will need to extend this the same way Katello will for the subscription uuid. @ezr-ondrej keep me honest here please.

Reply:
also @stejskalleos don't we want to extract the host finding/initialization to a separate methods for plugins being able to easily extend this part? next two lines are same for any other flow, but in case of sub-man or puppet, we may want to easily search by other identifier (cert name, uuid)

Reply:
Yes, there is an open PR for it: https://github.com/Katello/katello/pull/8965
You can override `prepare_host` method in plugin.

Reply:
but if you override the whole `prepare_host`, you'll need to repeat this part
```ruby
        @host.assign_attributes(host_params('host'))
        @host.save!
```

which seems to be the same in any flow. If you extract just 
```ruby
@host = Host.find_or_initialize_by(name: host_params('host')['name'])
```

to external method, it's easy to override just the host initialization. That would be more DRY I believe